codegen/winterfell/src/air/graph.rs

```diff
@@ -100,30 +100,69 @@ impl Codegen for Operation {
-            }
-            Operation::Sub(l_idx, r_idx) => {
-                let lhs = l_idx.to_string(ir, elem_type);
-                let rhs = r_idx.to_string(ir, elem_type);
-
-                format!("{lhs} - ({rhs})")
-            }
-            Operation::Mul(l_idx, r_idx) => {
-                let lhs = l_idx.to_string(ir, elem_type);
-                let rhs = r_idx.to_string(ir, elem_type);
-                format!("({lhs}) * ({rhs})")
-            }
+            Operation::Add(l_idx, r_idx) => binary_op_to_string(l_idx, r_idx, ir, '+', elem_type),
+            Operation::Sub(l_idx, r_idx) => binary_op_to_string(l_idx, r_idx, ir, '-', elem_type),
+            Operation::Mul(l_idx, r_idx) => binary_op_to_string(l_idx, r_idx, ir, '*', elem_type),
             Operation::Exp(l_idx, r_idx) => {
                 let lhs = l_idx.to_string(ir, elem_type);
+                let lhs = if is_leaf(l_idx, ir) {
+                    lhs
+                } else {
+                    format!("({lhs})")
+                };
                 match elem_type {
-                    ElemType::Base => format!("({lhs}).exp(Felt::new({r_idx}))"),
-                    ElemType::Ext => format!("({lhs}).exp(E::PositiveInteger::from({r_idx}_u64))"),
+                    ElemType::Base => format!("{lhs}.exp(Felt::new({r_idx}))"),
+                    ElemType::Ext => format!("{lhs}.exp(E::PositiveInteger::from({r_idx}_u64))"),
                 }
             }
         }
     }
 }
+
+/// Returns true if the operation at the specified node index is a leaf node in the constraint graph.
+fn is_leaf(idx: &NodeIndex, ir: &AirIR) -> bool {
+    !matches!(
+        ir.constraint_graph().node(idx).op(),
+        Operation::Add(_, _) | Operation::Sub(_, _) | Operation::Mul(_, _) | Operation::Exp(_, _)
+    )
+}
+
+/// Returns a string representation of a binary operation.
+fn binary_op_to_string(
+    l_idx: &NodeIndex,
+    r_idx: &NodeIndex,
+    ir: &AirIR,
+    binary_op: char,
+    elem_type: ElemType,
+) -> String {
+    match binary_op {
+        '+' => {
+            let lhs = l_idx.to_string(ir, elem_type);
+            let rhs = r_idx.to_string(ir, elem_type);
+            format!("{lhs} + {rhs}")
+        }
+        '-' => {
+            let lhs = l_idx.to_string(ir, elem_type);
+            let rhs = if is_leaf(r_idx, ir) {
+                r_idx.to_string(ir, elem_type)
+            } else {
+                format!("({})", r_idx.to_string(ir, elem_type))
+            };
+            format!("{lhs} - {rhs}")
+        }
+        '*' => {
+            let lhs = if is_leaf(l_idx, ir) {
+                l_idx.to_string(ir, elem_type)
+            } else {
+                format!("({})", l_idx.to_string(ir, elem_type))
+            };
+            let rhs = if is_leaf(r_idx, ir) {
+                r_idx.to_string(ir, elem_type)
+            } else {
+                format!("({})", r_idx.to_string(ir, elem_type))
+            };
```

Comment:
nit: similar to above - on the rhs we only care about `Operation::Add` and `Operation::Sub`. I think we just don't have any examples of e.g. 2 multiplications in a row, so there aren't examples of the redundancy.

Reply:
We had an example of two simultaneous multiplications in bitwise constraints. I missed it before. Thanks for catching this. I've added a fix but that doesn't reduce the redundancy here. I can see that there should be a better solution so I'll try to update the PR soon.

To reduce the redundancy, I'm assuming you had something like this in mind?:

```
/// Returns a string representation of a binary operation.
fn binary_op_to_string(
    ir: &AirIR,
    l_idx: &NodeIndex,
    r_idx: &NodeIndex,
    binary_op: char,
    elem_type: ElemType,
) -> String {
    let lhs = match binary_op {
        '+' | '-' => l_idx.to_string(ir, elem_type),
        '*' => {
            if is_leaf(l_idx, ir) {
                l_idx.to_string(ir, elem_type)
            } else {
                format!("({})", l_idx.to_string(ir, elem_type))
            }
        }
        _ => panic!("Unsupported binary operation."),
    };
    let rhs = match binary_op {
        '+' => r_idx.to_string(ir, elem_type),
        '-' | '*' => {
            if is_leaf(r_idx, ir) {
                r_idx.to_string(ir, elem_type)
            } else {
                format!("({})", r_idx.to_string(ir, elem_type))
            }
        }
        _ => panic!("Unsupported binary operation."),
    };
    format!("{lhs} {binary_op} {rhs}")
}
```

But incorporating the change for multiple multiplications in a row probably makes the previous approach make more sense.

Reply:
Nvm, I realize you had something else in mind.